hail/src/main/scala/is/hail/expr/Parser.scala

```diff
@@ -83,41 +84,39 @@ object Parser extends JavaTokenParsers {
     }
   }
 
-  def oneOfLiteral(s: String*): Parser[String] = oneOfLiteral(s.toArray)
-
   def oneOfLiteral(a: Array[String]): Parser[String] = new Parser[String] {
-    var hasEnd: Boolean = false
+    val root = ParseTrieNode.generate(a)
 
-    val m = a.flatMap { s =>
-      val l = s.length
-      if (l == 0) {
-        hasEnd = true
-        None
-      }
-      else if (l == 1) {
-        Some((s.charAt(0), ""))
-      }
-      else
-        Some((s.charAt(0), s.substring(1)))
-    }.groupBy(_._1).mapValues { v => oneOfLiteral(v.map(_._2)) }
+    val sb: StringBuilder = new StringBuilder()
```

Comment:
Another potential optimization: to avoid allocating in the parser (and to effectively intern the literal strings), the `hasEnd` flag on the trie nodes could be replaced by a possibly null `value: String`, which is a reference to the element of `a` that led to that node. Then the parser can just return `value`, or failure if null.

Reply:
ah, that's nice.